f/connectors/alerts/alerts_gcs.py

```diff
@@ -370,16 +426,25 @@ def prepare_alerts_metadata(alerts_metadata, territory_id):
 
     # Hash each row into a unique UUID; this will be used as the primary key for the metadata table
     # The hash is based on the most important columns for the metadata table, so that changes in other columns do not affect the hash
+    # The hash excludes data_source to prevent duplicate alerts when switching providers
     filtered_df["_id"] = pd.util.hash_pandas_object(
         filtered_df[
             ["territory_id", "month", "year", "description_alerts", "confidence"]
         ].sort_index(axis=1),
         index=False,
     )
 
-    # TODO: Currently, this script is only used for Terras alerts. Let's discuss a more sustainable approach with the alerts provider(s).
-    # Also, if this changes for future alerts, we will need to ensure that existing records are not overwritten.
-    filtered_df["data_source"] = "terras"
+    # Preserve existing data_source values for existing records
+    existing_data_sources = _get_existing_data_sources(
+        db_conn_info, table_name, filtered_df["_id"]
+    )
+
+    # Set data_source: preserve existing values, use new provider only for truly new records
+    def get_data_source(row):
+        record_id = str(row["_id"])
+        return existing_data_sources.get(record_id, alerts_provider)
```

Comment:
This isn't quite the same as "new rows will get `alerts_provider`, old rows remain unchanged". Rather it's "replace all false-y values with `alerts_provider`".

So the difference is in the (unexpected) case that there is an old row with null value.

It's fine in this case but it does have a smell like we are doing something wrong.

See my top-level comment for bigger-picture thoughts.